### PR TITLE
feat: Convert nested function to closure assist

### LINKED ISSUE #13230: Feature request: "convert function to closure" assist
It would be nice if I could easily turn a function that is declared inside another function / method into a closure (to enable further refactorings like removing a parameter because it can be captured instead).

### PR DESCRIPTION
**Pull request by jplatte**
Continuation of / closes #13467.
Resolves #13230.

r? @Veykril

### CODE REVIEW

File: crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+pub(crate) fn convert_nested_function_to_closure(
+    acc: &mut Assists,
+    ctx: &AssistContext<'_>,
+) -> Option<()> {
+    let name = ctx.find_node_at_offset::<ast::Name>()?;
+    let function = name.syntax().parent().and_then(ast::Fn::cast)?;
+
+    if !is_nested_function(&function) || is_generic(&function) {
+        return None;
+    }
```
Comment (lowr) [inline crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs:39]: Very much likely to be rare occasion, but I suppose we don't want to trigger this assist when the function has modifiers i.e. `async`, `const`, and/or `unsafe`. For unsafe, We could make the closure body an unsafe block though.

File: crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            let has_semicolon = has_semicolon(&function);
+            let params_text = params.syntax().text().to_string();
+            let params_text_trimmed =
+                params_text.strip_prefix("(").and_then(|p| p.strip_suffix(")"));
+
+            if let Some(closure_params) = params_text_trimmed {
+                let body = body.to_string();
+                let body = if has_semicolon { body } else { format!("{};", body) };
+                edit.replace(target, format!("let {} = |{}| {}", name, closure_params, body));
+            }
```
Comment (lowr) [inline crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs:60]: When `params_text_trimmed` is `None` IDE will show the assist without any edit attached. Can you move this condition outside `acc.add()`?

File: crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs (hunk 1 above)
Comment (Veykril) [reply crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs:39]: > For unsafe, We could make the closure body an unsafe block though.

That doesn't have the same semantics though, so I would say not to. (Calling the closure would be safe, opposed to the function which would be unsafe to call)

Comment (lowr) [reply crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs:39]: Right! I have no idea why I thought that was okay :/

File: crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+/// Returns whether the given function is nested within the body of another function.
+fn is_nested_function(function: &ast::Fn) -> bool {
+    function
+        .syntax()
+        .ancestors()
+        .skip(1)
+        .find_map(ast::Item::cast)
+        .map_or(false, |it| matches!(it, ast::Item::Fn(_)))
```
[the remark this answers is not in the document]
Comment (jplatte) [reply crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs:72]: I reckon it's extremely uncommon to have a function nested inside a const or static initializer, but sure why not 😄

[the remark this answers is not in the document]
Comment (Veykril) [reply crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs]: I'm not going to deny that, but you never know (If we want to be very pedantic, we should actually check for enum variants, const arguments and const parameter defaults as well 😶)

**Comment by Veykril** [comment]
Thanks, to you too of course @mdx97!

File: crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jplatte) [reply crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs]: I understand const arguments and const parameter defaults, but enum variants? You mean enum variant numbers, as in you can have a block expression in `enum Foo { Bar = <here> }`? 😂

**Pull request merged**